### DISCUSSION TITLE
Failed to identify PCI_SLOT_NAME=0000:c1:00.0: failed to issue admin command 0x06: 83eba1fa14bdefee1f6b4bcf06069b298e1c338f [NVMe SSD Controller PM174X] has not been opened

### DISCUSSION THREAD

**Discussion by xiaochunlee** [fwupd/fwupd#5841] [Q&A] [answered]
Hi, 
There is a system running RHEL9.0 with PCIE gen5 and U.3 NVMe , when the genome start, the fwupd will show the below messages. And this error will result in nvme hot plug problems, it indirectly get /dev/nvmeX (such as /dev/nvme4) device reference count in Linux kernel, and doesn't put it until the fwup daemon deactivated. Since the reference count couldn't be release in time, nvme controller instance number will not be released in linux kernel after unplugging the nvme4 in the second slot. Note, only the nvme4 (the second slot) has this issue, the other NVMes didn't see this symptoms.

Tracing the fwupdmgr code, it seems open the /dev/nvme4 dev node failed, but this dev node is there before the fwupd is running and the smartd can also open it.  

So is there anyone knows what happened in the systems? I was hoping your suggestions and helps, really appreciate it.
-->>
May 19 18:53:51 localhost.localdomain systemd[1]: Starting Firmware update daemon...
May 19 18:53:52 localhost.localdomain fwupd[5101]: 10:53:52:0042 FuPluginUpower       failed to query power type
May 19 18:53:52 localhost.localdomain fwupd[5101]: 10:53:52:0188 FuEngine             failed to add device /sys/devices/pci0000:20/0000:20:01.1/0000:21:00.0: EEPROM size invalid, got 0x80000, expect
ed 0x40000
May 19 18:53:52 localhost.localdomain fwupd[5101]: 10:53:52:0551 FuEngine             failed to add device /sys/devices/pci0000:c0/0000:c0:01.1/0000:c1:00.0/nvme/nvme4: failed to identify PCI_SLOT_NAME=0000:c1:00.0: failed to issue admin command 0x06: 83eba1fa14bdefee1f6b4bcf06069b298e1c338f [NVMe SSD Controller PM174X] has not been opened
May 19 18:53:52 localhost.localdomain systemd[1]: Started Firmware update daemon.
May 19 18:54:58 localhost.localdomain fwupd[5101]: 10:54:58:0203 FuEngine             failed to add device /sys/devices/pci0000:c0/0000:c0:01.1/0000:c1:00.0/nvme/nvme10: failed to open /dev/nvme10:Resource temporarily unavailable
<<--

**Comment by hughsie** [comment]
>  failed to add device /sys/devices/pci0000:20/0000:20:01.1/0000:21:00.0: EEPROM size invalid

This look unrelated to the NVMe device -- correct? It would be interesting to know what this device is.

> NVMe SSD Controller PM174X] has not been opened

Hmm, this is strange. What fwupd version please?

**Reply by xiaochunlee** [reply]
Thanks your quick responds. Device 0000:21:00.0 is a Broadcom NIC adapter, the "lspci -vs 21:00.0" shown as below.  
-->>
[root@localhost ~]# lspci -vs 21:00.0
21:00.0 Ethernet controller: Broadcom Inc. and subsidiaries NetXtreme BCM5719 Gigabit Ethernet PCIe (rev 01)
        Subsystem: Lenovo Device 4104
        Physical Slot: 13
        Flags: bus master, fast devsel, latency 0, IRQ 108, NUMA node 0, IOMMU group 12
        Memory at 70480fb0000 (64-bit, prefetchable) [size=64K]
        Memory at 70480fa0000 (64-bit, prefetchable) [size=64K]
        Memory at 70480f90000 (64-bit, prefetchable) [size=64K]
        Expansion ROM at c48c0000 [disabled] [size=256K]
        Capabilities: [48] Power Management version 3
        Capabilities: [50] Vital Product Data
        Capabilities: [58] MSI: Enable- Count=1/8 Maskable- 64bit+
        Capabilities: [a0] MSI-X: Enable+ Count=17 Masked-
        Capabilities: [ac] Express Endpoint, MSI 00
        Capabilities: [100] Advanced Error Reporting
        Capabilities: [13c] Device Serial Number 00-00-e4-3d-1a-61-53-e2
        Capabilities: [150] Power Budgeting <?>
        Capabilities: [160] Virtual Channel
        Kernel driver in use: tg3
        Kernel modules: tg3
<--
The fwupd version I caught as below, I am not sure it is you wanted. it was integrated in RHEL9.0 GA, no any updates here.
-->>
[root@localhost ~]# fwupdmgr --version
client version: 1.7.4
compile-time dependency versions
        gusb:   0.3.8

daemon version: 1.7.4
[root@localhost ~]# 
<<--

**Comment by xiaochunlee** [comment]
Here is the fwupd unit output by command "journalctl -u fwupd" after login GUI

-->>
May 22 15:01:50 localhost.localdomain systemd[1]: Starting Firmware update daemon...
May 22 15:01:50 localhost.localdomain fwupd[8979]: 07:01:50:0419 FuPluginUpower       failed to query power type
May 22 15:01:50 localhost.localdomain fwupd[8979]: 07:01:50:0659 FuEngine             failed to add device /sys/devices/pci0000:20/0000:20:01.1/0000:21:00.0: EEPROM size invalid, got 0x80000, expected 0x40000
May 22 15:01:51 localhost.localdomain fwupd[8979]: 07:01:51:0265 FuEngine             failed to add device /sys/devices/pci0000:c0/0000:c0:01.1/0000:c1:00.0/nvme/nvme4: failed to identify PCI_SLOT_NAME=0000:c1:00.0: failed to issue admi
n command 0x06: 83eba1fa14bdefee1f6b4bcf06069b298e1c338f [NVMe SSD Controller PM174X] has not been opened
May 22 15:01:51 localhost.localdomain systemd[1]: Started Firmware update daemon.
<<--

And then unplug the the second solt nvme,  the command "journalctl -u fwupd" shown no any changes. But the command 
"ll /sys/class/nvme-subsystem/nvme-subsys4/" listed as below, the controller instance nvme4 under /sys/class/nvme-subsystem/nvme-subsys4/ should be released after unplug, but it was there with an unavailable soft links, because it points link file "../../../pci0000:c0/0000:c0:01.1/0000:c1:00.0/nvme/nvme4" was destroyed. Actually, after unplugging nvme that in the second slot, the corresponding file "/sys/class/nvme-subsystem/nvme-subsys4/" should be removed, but it doesn't.
-->>
[root@localhost ~]# ll /sys/class/nvme-subsystem/nvme-subsys4/
total 0
-r--r--r--. 1 root root 4096 May 22  2023 firmware_rev
-rw-r--r--. 1 root root 4096 May 22 15:31 iopolicy
-r--r--r--. 1 root root 4096 May 22  2023 model
lrwxrwxrwx. 1 root root    0 May 22 15:19 nvme4 -> ../../../pci0000:c0/0000:c0:01.1/0000:c1:00.0/nvme/nvme4
drwxr-xr-x. 2 root root    0 May 22 15:31 power
-r--r--r--. 1 root root 4096 May 22  2023 serial
-r--r--r--. 1 root root 4096 May 22 15:19 subsysnqn
lrwxrwxrwx. 1 root root    0 May 22  2023 subsystem -> ../../../../class/nvme-subsystem
-r--r--r--. 1 root root 4096 May 22 15:31 subsystype
-rw-r--r--. 1 root root 4096 May 22  2023 uevent
<<--

The next I plugged the unplug nvme in the second slot. The command "journalctl -u fwupd" appended one line listed as below.
-->>
[root@localhost ~]# journalctl -u fwupd >fwupd.log 
[root@localhost ~]# vim fwupd.log 
May 22 15:01:50 localhost.localdomain systemd[1]: Starting Firmware update daemon...
May 22 15:01:50 localhost.localdomain fwupd[8979]: 07:01:50:0419 FuPluginUpower       failed to query power type
May 22 15:01:50 localhost.localdomain fwupd[8979]: 07:01:50:0659 FuEngine             failed to add device /sys/devi
ces/pci0000:20/0000:20:01.1/0000:21:00.0: EEPROM size invalid, got 0x80000, expected 0x40000
May 22 15:01:51 localhost.localdomain fwupd[8979]: 07:01:51:0265 FuEngine             failed to add device /sys/devi
ces/pci0000:c0/0000:c0:01.1/0000:c1:00.0/nvme/nvme4: failed to identify PCI_SLOT_NAME=0000:c1:00.0: failed to issue
admin command 0x06: 83eba1fa14bdefee1f6b4bcf06069b298e1c338f [NVMe SSD Controller PM174X] has not been opened
May 22 15:01:51 localhost.localdomain systemd[1]: Started Firmware update daemon.
May 22 15:49:51 localhost.localdomain fwupd[8979]: 07:49:51:0018 FuEngine             failed to add device /sys/devi
ces/pci0000:c0/0000:c0:01.1/0000:c1:00.0/nvme/nvme10: failed to open /dev/nvme10: Resource temporarily unavailable
<<--

And since the nvme controller instance number 4 wasn't released when unplug, so when I plug it, the minimum available controller instance  number in the system is 10, so it become 10 for the original nvme.  The command "ll /sys/class/nvme-subsystem/nvme-subsys4/" output listed as following.
-->>
[root@localhost ~]# ll /sys/class/nvme-subsystem/nvme-subsys4/
total 0
-r--r--r--. 1 root root 4096 May 22  2023 firmware_rev
-rw-r--r--. 1 root root 4096 May 22 15:31 iopolicy
-r--r--r--. 1 root root 4096 May 22  2023 model
drwxr-xr-x. 3 root root    0 May 22 15:49 ng4n1
lrwxrwxrwx. 1 root root    0 May 22 15:55 nvme10 -> ../../../pci0000:c0/0000:c0:01.1/0000:c1:00.0/nvme/nvme10
lrwxrwxrwx. 1 root root    0 May 22 15:19 nvme4 -> ../../../pci0000:c0/0000:c0:01.1/0000:c1:00.0/nvme/nvme4
drwxr-xr-x. 8 root root    0 May 22 15:49 nvme4n1
drwxr-xr-x. 2 root root    0 May 22 15:31 power
-r--r--r--. 1 root root 4096 May 22  2023 serial
-r--r--r--. 1 root root 4096 May 22 15:19 subsysnqn
lrwxrwxrwx. 1 root root    0 May 22  2023 subsystem -> ../../../../class/nvme-subsystem
-r--r--r--. 1 root root 4096 May 22 15:31 subsystype
-rw-r--r--. 1 root root 4096 May 22  2023 uevent
[root@localhost ~]# 
<<--

Since the original nvme controller number wasn't released, the smartd will show error  "Device: /dev/nvme4, open() of NVMe device failed: No such device" and smartd doesn't monitor the new named nvme(nvme10). 
For the same reason, other tools like command "nvme --list" also show error like "Failed to open /sys/class/nvme-subsystem/nvme-subsys4/nvme4: No such file or directory"

Thus, from the observation and analysis, the root cause is fwupd failed to open nvme controller (nvme4), Howerver the nvme4 works well, mount /dev/nvme4n1, write files in it, it works. Regarding why fwupd failed to open nvme controller, it need your help, many thanks!

**Reply by hughsie** [reply]
> 1.7.4

This is super old -- can you try a fwupd 1.8.10 release from RHEL 9.2? If that still fails then we can debug some more. Thanks.

**Reply by xiaochunlee** [reply]
OK，I'll try it and let you know the results. 
And here is some updates
When I run "systemctl stop fwupd" , the nvme4 reference will be released by Linux kernel and finally the subsystem files will be remove from OS.

**Reply by xiaochunlee** [reply]
> > 1.7.4
> 
> This is super old -- can you try a fwupd 1.8.10 release from RHEL 9.2? If that still fails then we can debug some more. Thanks.

The fwupd version comes to 1.8.10, but the symptom is still there. More info listed as below. I just updated fwupd and fwupd-plugin-flashrom  to version 1.8.10-2.el9, did install all RHEL9.2.
-->>
[root@localhost ~]# fwupdmgr --version
compile   org.freedesktop.fwupd         1.8.10
compile   com.hughsie.libjcat           0.1.6
runtime   org.freedesktop.fwupd-efi     1.8.10
compile   org.freedesktop.gusb          0.3.8
runtime   com.dell.libsmbios            2.4
runtime   org.freedesktop.gusb          0.3.8
runtime   org.freedesktop.fwupd         1.8.10
runtime   org.kernel                    5.14.0-004-nvme
[root@localhost ~]# 
<<--

Command "journalctl -u fwupd.service" shows the same errors as before.
-->>
May 23 11:06:03 localhost.localdomain systemd[1]: Starting Firmware update daemon...
May 23 11:06:09 localhost.localdomain fwupd[10541]: 03:06:09.288 FuEngine             failed to add device /sys/devi
ces/pci0000:20/0000:20:01.1/0000:21:00.0: EEPROM size invalid, got 0x80000, expected 0x40000
May 23 11:06:09 localhost.localdomain fwupd[10541]: 03:06:09.709 FuEngine             failed to add device /sys/devi
ces/pci0000:c0/0000:c0:01.1/0000:c1:00.0/nvme/nvme4: failed to identify PCI_SLOT_NAME=0000:c1:00.0: failed to issue
admin command 0x06: 83eba1fa14bdefee1f6b4bcf06069b298e1c338f [NVMe SSD Controller PM174X] has not been opened
May 23 11:06:39 localhost.localdomain systemd[1]: Started Firmware update daemon.
~                                                                               
<<--

**Comment by xiaochunlee** [comment]
I checked fwupd code by clone from here, In main branch, I found that  there are partial code as "if (priv->fd == 0) {" at line 1601 in function fu_udev_device_ioctl(), so does it treat zero fd as a not opened? Actually, in this case, the fwupd invoke open and return a minimum file descriptor here, it's zero indeed.
The partial strace log listed as following, it shows the file descriptor zero was closed. 
-->
2039790 openat(17, "devices", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 16
2039791 newfstatat(16, "", {st_mode=S_IFDIR|0755, st_size=0, ...}, AT_EMPTY_PATH) = 0
2039792 close(17)                               = 0
2039793 close(16)                               = 0
2039794 access("/sys/devices/uevent", F_OK)     = -1 ENOENT (No such file or directory)
2039795 close(0)                                = 0
2039796 access("/sys/devices/pci0000:20/0000:20:01.1/0000:21:00.1/rom", F_OK) = 0
2039797 newfstatat(AT_FDCWD, "/sys/devices/pci0000:20/0000:20:01.1/vendor", {st_mode=S_IFREG|0444, st_size=4096, ...
        }, AT_SYMLINK_NOFOLLOW) = 0
2039798 openat(AT_FDCWD, "/sys/devices/pci0000:20/0000:20:01.1/vendor", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 0
2039799 newfstatat(0, "", {st_mode=S_IFREG|0444, st_size=4096, ...}, AT_EMPTY_PATH) = 0
2039800 read(0, "0x1022\n", 4104)               = 7
2039801 close(0)                                = 0

...

2044387 openat(AT_FDCWD, "/run/udev/data/+pci:0000:c1:00.0", O_RDONLY|O_CLOEXEC) = 0
2044388 newfstatat(0, "", {st_mode=S_IFREG|0644, st_size=283, ...}, AT_EMPTY_PATH) = 0
2044389 newfstatat(0, "", {st_mode=S_IFREG|0644, st_size=283, ...}, AT_EMPTY_PATH) = 0
2044390 read(0, "I:2477002128\nE:ID_PCI_CLASS_FROM"..., 4096) = 283
2044391 read(0, "", 4096)                       = 0
2044392 close(0)                                = 0
2044393 openat(AT_FDCWD, "/dev/nvme11", O_RDONLY) = 0
2044394 write(2, "\33[32m08:32:19.838 ", 18)    = 18
2044395 write(2, "FuEngine             ", 21)   = 21
2044396 write(2, "\33[31mfailed to add device /sys/d"..., 266) = 266
2044397 newfstatat(AT_FDCWD, "/sys/devices/pci0000:c0/0000:c0:01.1/0000:c1:00.0/nvme/nvme11/nvme4c11n1/vendor", 0x7f
        fc6b770370, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
<<--

**Reply by hughsie** [reply]
So really the fault is in fwupd that assumes that zero is not a valid file descriptor?

**Reply by hughsie** [reply]
Does the wip/hughsie/fd-nonzero branch work like you'd expect? Thanks.

**Reply by xiaochunlee** [reply]
Yes, I'll try it. Many thanks!
But I stuck in the first step when I tried to compile the project. I following the instructions to get the Dockerfile at https://github.com/fwupd/fwupd/tree/main/contrib, but there isn't file generate_docker.py under contrib.
I also tried to meson build directly, but as you know, it needs many dependencies tools, I was stopping in finding "gobject-introspection-1.0", it was needed by building libxmlb.
Is there any suggestions for us to build this project? Thanks!

**Reply by hughsie** [reply]
Would a Fedora or RHEL package be appreciated? Fedora is easy, RHEL might take a few more hours.

**Reply by xiaochunlee** [reply]
I test it on RHEL9.0, so I prefer RHEL. but I don't know whether Fedora packages can install on RHEL.

**Reply by xiaochunlee** [reply]
Hello, hughsie
I tried to build rpm packages for Fedora in docker, but it pop out errors listed as following. what should I do for this? Appreciate it!
-->>
32/32 fu-self-test                           OK               9.50s

Summary of Failures:

31/32 redfish-self-test              FAIL             1.76s   killed by signal 6 SIGABRT

Ok:                 31  
Expected Fail:      0   
Fail:               1   
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0   

Full log written to /github/workspace/build/meson-private/dist-build/meson-logs/testlog.txt
FAILED: meson-internal__test 
/usr/bin/meson test --no-rebuild --print-errorlogs
ninja: build stopped: subcommand failed.
Running unit tests on the distribution package failed
Dist check build directory was /github/workspace/build/meson-private/dist-build
FAILED: meson-internal__dist 
/usr/bin/meson dist
ninja: build stopped: subcommand failed.
[root@b61a368e3750 workspace]# 
<<--

**Reply by hughsie** [reply, marked as answer]
Can you try these packages please: https://people.redhat.com/rhughes/packages/fwupd/el9/

**Reply by xiaochunlee** [reply]
Yeah, I'll, thanks!

**Reply by xiaochunlee** [reply]
Yes, I try this new rpm packages, it works well, and the nvme controller instance number can be released after unplug the device. Thanks!
How about I file a new bug on RedHat bugzilla to track this issue. I was just wondering if RedHat could write a KB for this, it would be better for our mutual customers who has this issues. Thanks!

**Reply by hughsie** [reply]
I think Lenovo already did that :) https://bugzilla.redhat.com/show_bug.cgi?id=2209944

Thanks for testing; I'll make sure that this fix gets included in the next RHEL fwupd rebase.

**Reply by xiaochunlee** [reply]
Yes, I saw that! thanks!
